src/Shared/DataSourceHelper/DataSourceHelper.ts

```diff
@@ -0,0 +1,170 @@
+    _datasource: {
+        _constraints: OfflineConstraint[] | string;
+        _entity: string;
+        _sorting: string[][];
+    };
+    datasource: {
+        type: "microflow" | "entityPath" | "database" | "xpath";
+    };
+    update: (obj: mendix.lib.MxObject | null, callback?: () => void) => void;
+    sequence: (sequence: string[], callback?: () => void) => void;
+    _entity: string;
+    __customWidgetDataSourceHelper?: DataSourceHelper;
+}
+
+export class DataSourceHelper {
+    // The version of a Datasource is static, it never changes.
+    static VERSION: Version = { major: 1, minor: 0, patch: 0 };
+    // Expose the version dataSourceHelper instances.
+    public version: Version = DataSourceHelper.VERSION;
+    private delay = 50;
+    private timeoutHandle?: number;
+    private store: ConstraintStore = { constraints: {}, sorting: {} };
+    private widget: ListView;
+    private updateInProgress = false;
+    private requiresUpdate = false;
+    private widgetVersionRegister: {
+        [version: string]: string[];
+    } = {};
+
+    constructor(widget: ListView) {
+        this.compatibilityCheck(widget);
+        this.widget = widget;
+    }
+
+    static getInstance(widget: ListView, widgetId: string, version: Version) {
+        if (!widget.__customWidgetDataSourceHelper) {
+            widget.__customWidgetDataSourceHelper = new DataSourceHelper(widget);
+        }
+        widget.__customWidgetDataSourceHelper.versionCompatibility(version, widgetId);
+
+        return widget.__customWidgetDataSourceHelper;
+    }
+
+    setSorting(widgetId: string, sortConstraint: string[]) {
+        this.store.sorting[widgetId] = sortConstraint;
+        this.registerUpdate();
+    }
+
+    setConstraint(widgetId: string, constraint: string | OfflineConstraint) {
+        this.store.constraints[widgetId] = constraint as string | OfflineConstraint;
+        this.registerUpdate();
+    }
+
+    private registerUpdate() {
+        if (this.timeoutHandle) {
+            window.clearTimeout(this.timeoutHandle);
+        }
+        if (!this.updateInProgress) {
+            this.timeoutHandle = window.setTimeout(() => {
+                this.updateInProgress = true;
+                // TODO Check if there's currently no update happening on the listView coming from another
+                // Feature/functionality/widget which does not use DataSourceHelper
+                this.iterativeUpdateDataSource();
+          }, this.delay);
+        } else {
+            this.requiresUpdate = true;
+        }
+    }
+
+    private iterativeUpdateDataSource() {
+        this.updateDataSource(() => {
+            if (this.requiresUpdate) {
+                this.requiresUpdate = false;
+                this.iterativeUpdateDataSource();
+            } else {
+                this.updateInProgress = false;
+            }
+        });
+    }
+
+    private versionCompatibility(version: Version, widgetId: string) {
```

Comment:
Can we simplify this, by just checking the  version, there is low/no change that some other widget will do the  same thing....